app/actors/ProxyConnection.java

```diff
@@ -0,0 +1,231 @@
+        while (!_originatorMessageQueue.isEmpty()) {
+            _destinationClient.send(_originatorMessageQueue.remove());
+        }
+
+        // Send any buffered messages from the destination to the orignator
+        _destinationClient.establishProxy();
+    }
+
+    private final MetricsFactory _metricsFactory;
+    private final AtomicBoolean _isProxied = new AtomicBoolean(false);
+    private final Queue<String> _originatorMessageQueue = new LinkedList<>();
+    private WebSocket.In<String> _originatorIn;
+    private WebSocket.Out<String> _originatorOut;
+    private ProxyWebSocketClient _destinationClient;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProxyConnection.class);
+
+    private static final class DestinationConnected {}
+
+    private static final class OriginatorMessage {
+        OriginatorMessage(final String message) {
+            _message = message;
+        }
+
+        public String getMessage() {
+            return _message;
+        }
+
+        private final String _message;
+    }
+
+    private final class ProxyWebSocketClient extends WebSocketClient {
+
+        ProxyWebSocketClient(final URI uri) {
+            super(uri);
+        }
+
+        @Override
+        public void onOpen(final ServerHandshake handshakedata) {
+            ProxyConnection.this.getSelf().tell(new DestinationConnected(), ProxyConnection.this.getSelf());
+        }
+
+        @Override
+        public void onMessage(final String message) {
+            // Unsure what the underlying threading model in WebSocketClient is; so play it safe
+            synchronized (_isProxied) {
+                if (_isProxied.get()) {
+                    _originatorOut.write(message);
+                } else {
+                    _destinationMessageQueue.add(message);
+                }
+            }
+        }
+
+        @Override
+        public void onClose(final int code, final String reason, final boolean remote) {
+            LOGGER.info()
+                    .setMessage("Destination connection closed")
+                    .addData("actor", self())
+                    .addData("destination", getURI())
+                    .log();
+            ProxyConnection.this.getSelf().tell(PoisonPill.getInstance(), ProxyConnection.this.getSelf());
+        }
+
+        @Override
+        public void onError(final Exception ex) {
+            LOGGER.warn()
+                    .setMessage("Destination connection error")
+                    .addData("actor", self())
+                    .addData("destination", getURI())
+                    .setThrowable(ex)
+                    .log();
+            ProxyConnection.this.getSelf().tell(PoisonPill.getInstance(), ProxyConnection.this.getSelf());
+        }
+
+        public void establishProxy() {
+            // Unsure what the underlying threading model in WebSocketClient is; so play it safe
+            synchronized (ProxyConnection.this._isProxied) {
```

Comment:
By implication of the above, with the queue in the actor the drain on establish is also in the actor now.